src/wkt/src/internal.rs

```diff
@@ -18,19 +18,78 @@
 //! for Rust. They are undocumented and may change at any time.
 
 pub struct F32;
+pub struct F64;
+
+macro_rules! impl_serialize_as {
+    ($t: ty, $ser_fn: ident) => {
+        fn serialize_as<S>(value: &$t, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: serde::ser::Serializer,
+        {
+            match value {
+                x if x.is_nan() => serializer.serialize_str("NaN"),
+                x if x.is_infinite() && x.is_sign_negative() => {
+                    serializer.serialize_str("-Infinity")
+                }
+                x if x.is_infinite() => serializer.serialize_str("Infinity"),
+                x => serializer.$ser_fn(*x),
+            }
+        }
+    };
+}
 
-impl serde_with::SerializeAs<f32> for F32 {
-    fn serialize_as<S>(value: &f32, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: serde::ser::Serializer,
-    {
-        match value {
-            x if x.is_nan() => serializer.serialize_str("NaN"),
-            x if x.is_infinite() && x.is_sign_negative() => serializer.serialize_str("-Infinity"),
-            x if x.is_infinite() => serializer.serialize_str("Infinity"),
-            x => serializer.serialize_f32(*x),
+macro_rules! impl_visitor {
+    ($name: ident, $t: ty) => {
+        struct $name;
+
+        impl serde::de::Visitor<'_> for $name {
+            type Value = $t;
+
+            fn visit_str<E>(self, value: &str) -> std::result::Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                // Handle special strings, see https://protobuf.dev/programming-guides/json/.
+                match value {
+                    "NaN" => Ok(<$t>::NAN),
+                    "Infinity" => Ok(<$t>::INFINITY),
+                    "-Infinity" => Ok(<$t>::NEG_INFINITY),
+                    _ => Err(serde::de::Error::invalid_value(
+                        serde::de::Unexpected::Other(value),
+                        &format!(
+                            "a valid ProtoJSON string for {} (NaN, Infinity, -Infinity)",
+                            std::any::type_name::<$t>()
+                        )
+                        .as_str(),
+                    )),
+                }
+            }
+
+            // Floats and doubles in serde_json are f64.
+            fn visit_f64<E>(self, value: f64) -> std::result::Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                // Casting f64 to f32 to produce the closest possible float value.
```

Comment:
nit: when `$t == f64` the comment is not quite right, maybe:

```suggestion
                // This is trivial for `f64`. For `f32`, casting f64 to f32 is guaranteed to produce the closest possible float value:
```

Reply:
Done.